timemachine/md/smc.py

```diff
@@ -19,144 +19,190 @@
     Returns
     -------
     trajs_dict
-        "sample_traj"
-            [K-1, N] list of snapshots
+        "traj"
+            [K-1, N] list of snapshots only if `store_intermediate_traj` = True.
+            [1, N] list of snapshots if `store_intermediate_traj` = False.
         "incremental_log_weights_traj"
             [K-1, N] array of incremental log weights
         "ancestry_traj"
             [K-1, N] array of ancestor idxs
         "log_weights_traj"
             [K, N] array of accumulated log weights
+        "lambdas_traj"
+            [K] array of adaptive lambdas
 
     References
     ----------
-    * Arnaud Doucet's annotated bibliography of SMC
-        https://www.stats.ox.ac.uk/~doucet/smc_resources.html
-    * [Dai, Heng, Jacob, Whiteley, 2020] An invitation to sequential Monte Carlo samplers
-        https://arxiv.org/abs/2007.11936
-
-    See Also
-    --------
-    * get_endstate_samples_from_smc_result
+    * [Zhou, Johansen, Aston, 2016]
+        Towards Automatic Model Comparison: An Adaptive Sequential Monte Carlo Approach
+        https://arxiv.org/pdf/1303.3123 (Algorithm #4)
     """
     n = len(samples)
-    log_weights: NDArray[np.float64] = np.zeros(n, dtype=np.float64)
+
+    log_weights: LogWeights = np.zeros(n)
+    norm_log_weights: LogWeights = log_weights - logsumexp(log_weights)
 
     # store
     sample_traj = [samples]
     ancestry_traj = [np.arange(n)]
     log_weights_traj = [np.array(log_weights)]
     incremental_log_weights_traj = []  # note: redundant but convenient
+    lambdas_traj = [0.0]
 
-    def accumulate_results(samples, indices, log_weights, incremental_log_weights):
-        sample_traj.append(samples)
+    def accumulate_results(samples, indices, log_weights, incremental_log_weights, lam_target):
+        if store_intermediate_traj:
+            sample_traj.append(samples)
+        else:
+            # only store one intermediate set of samples to reduce memory usage
+            sample_traj[0] = samples
         ancestry_traj.append(indices)
         log_weights_traj.append(np.array(log_weights))
         incremental_log_weights_traj.append(np.array(incremental_log_weights))
+        lambdas_traj.append(lam_target)
+
+    lam_initial: Lambda = 0.0
+    lam_target: Lambda = 1.0
+    current_iteration: Iteration = 0
 
-    # Note: loop bounds [1:-1] intentional:
-    #   these steps are the only ones that contribute to the SMC result,
-    #   and some care is needed to extract endstate samples at t=0, t=T-1
-    #   See also
-    #   * discussion at https://github.com/proteneer/timemachine/pull/718#discussion_r854276326
-    #   * helper function get_endstate_samples_from_smc_result
-    for lam_initial, lam_target in zip(lambdas[:-2], lambdas[1:-1]):
-        # update log weights
-        incremental_log_weights = log_prob(sample_traj[-1], lam_target) - log_prob(sample_traj[-1], lam_initial)
-        log_weights += incremental_log_weights
+    while True:
```

Comment:
Preference to avoid potentially unbounded loops: consider replacing with `for _ in range(max_num_lambdas): ` or similar